src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java

```diff
@@ -0,0 +1,955 @@
+                LOGGER.error("Received a modded custom payload packet from a server with an unknown or not accepted channel. Disconnecting server.");
+                throw new IllegalStateException("A server send a packet with an unknown or not accepted channel, while negotiation succeeded. Somebody changed the channels known to NeoForge!");
+            }
+
+            registration.handle(
+                    packet.payload(),
+                    new ConfigurationPayloadContext(
+                            configurationPacketListener::send,
+                            new ClientPacketHandler(configurationPacketListener),
+                            (task) -> LOGGER.warn("Tried to finish a task on the client. This should not happen. Ignoring. Task: {}", task),
+                            new EventLoopSynchronizedWorkHandler<>(configurationPacketListener.getMainThreadEventLoop(), packet.payload()),
+                            PacketFlow.CLIENTBOUND,
+                            listener.getConnection().channel().pipeline().lastContext(),
+                            Optional.empty()));
+        } else if (listener instanceof ClientGamePacketListener playPacketListener) {
+            //Get the configuration channel for the packet.
+            final NetworkChannel channel = payloadSetup.play().get(packet.payload().id());
+
+            //Check if the channel should even be processed.
+            if (channel == null) {
+                LOGGER.warn("Received a modded custom payload packet from a server with an unknown or not accepted channel. Disconnecting server.");
+                listener.getConnection().disconnect(Component.translatable("multiplayer.disconnect.incompatible", "NeoForge %s".formatted(NeoForgeVersion.getVersion())));
+                return;
+            }
+
+            //We are in the play phase, so lookup packet listeners for that
+            final PlayRegistration<?> registration = knownPlayRegistrations.get(channel.id());
+            if (registration == null) {
+                LOGGER.error("Received a modded custom payload packet from a server with an unknown or not accepted channel. Disconnecting server.");
+                throw new IllegalStateException("A server send a packet with an unknown or not accepted channel, while negotiation succeeded. Somebody changed the channels known to NeoForge!");
+            }
+
+            registration.handle(
+                    packet.payload(),
+                    new PlayPayloadContext(
+                            playPacketListener::send,
+                            new ClientPacketHandler(playPacketListener),
+                            new EventLoopSynchronizedWorkHandler<>(playPacketListener.getMainThreadEventLoop(), packet.payload()),
+                            PacketFlow.CLIENTBOUND,
+                            listener.getConnection().channel().pipeline().lastContext(),
+                            Optional.empty()));
```

Comment:
Could we add an extension to `ClientGamePacketListener` that adds a `@Nullable Player getPlayer()` and implement that in `ClientCommonPacketListenerImpl` to return `this.minecraft.player` as it has a minecraft context?